File: c++/MinStack.cpp
class MinStack {
private:
    vector<int> q;
    vector<int> min;
public:
    void push(int x) {
        q.push_back(x);
        if(min.size()==0)
            min.push_back(x);
        else
        {
            int curMin = getMin();
            if(x<curMin)
                curMin = x;
            min.push_back(curMin);
        }
    }

    void pop() {
        q.pop_back();
        min.pop_back();
    }

    int top() {
        return q.back();
    }

    int getMin() {
        return min.back();
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(x);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */

File: python/min_stack.py
class MinStack(object):

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.p = []
        
    def push(self, x):
        """
        :type x: int
        :rtype: void
        """
        curMin = self.getMin()
        if curMin == None or x < curMin:
            curMin = x
        self.p.append((x, curMin))

    def pop(self):
        """
        :rtype: void
        """
        self.p.pop()

        

    def top(self):
        """
        :rtype: int
        """
        try:
            return self.p[-1][0]
        except IndexError:
            return None
    
        

    def getMin(self):
        """
        :rtype: int
        """
        try:
            return self.p[-1][1]
        except IndexError:
            return None
        


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(x)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()